File: rle_decompress.go
// rle_decompress.go
package main

import (
	"bufio"
	"errors"
	"flag"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

const (
	reset  = "\033[0m"
	green  = "\033[92m"
	red    = "\033[91m"
	yellow = "\033[93m"
)

func colorize(text, color string) string {
	return color + text + reset
}

const ESCAPE = '\\'

func decompress(text string) (string, error) {
	if text == "" {
		return "", nil
	}
	var result strings.Builder
	n := len(text)
	i := 0
	for i < n {
		ch := text[i]
		if ch == ESCAPE {
			if i+1 < n && text[i+1] == ESCAPE {
				result.WriteByte(ESCAPE)
				i += 2
				continue
			}
			if i+1 >= n {
				return "", errors.New("unexpected end after escape")
			}
			repeatChar := text[i+1]
			i += 2
			numStr := ""
			for i < n && text[i] >= '0' && text[i] <= '9' {
				numStr += string(text[i])
				i++
			}
			if numStr == "" {
				return "", errors.New("missing number after escape")
			}
			count, err := strconv.Atoi(numStr)
			if err != nil {
				return "", err
			}
			for k := 0; k < count; k++ {
				result.WriteByte(repeatChar)
			}
		} else {
			result.WriteByte(ch)
			i++
		}
	}
	return result.String(), nil
}

func readInput(filename string) (string, error) {
	if filename == "-" || filename == "" {
		reader := bufio.NewReader(os.Stdin)
		var builder strings.Builder
		for {
			line, err := reader.ReadString('\n')
			if err != nil && err != io.EOF {
				return "", err
			}
			builder.WriteString(line)
			if err == io.EOF {
				break
			}
		}
		return builder.String(), nil
	}
	data, err := os.ReadFile(filename)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func writeOutput(filename string, content string) error {
	if filename == "-" || filename == "" {
		fmt.Print(content)
		return nil
	}
	return os.WriteFile(filename, []byte(content), 0644)
}

func main() {
	verbose := flag.Bool("v", false, "show statistics")
	flag.Usage = func() {
		fmt.Println(colorize("Usage: rle_decompress <input> [output] [-v]", yellow))
	}
	flag.Parse()
	args := flag.Args()
	if len(args) < 1 {
		flag.Usage()
		os.Exit(1)
	}
	inputFile := args[0]
	outputFile := ""
	if len(args) >= 2 {
		outputFile = args[1]
	}

	data, err := readInput(inputFile)
	if err != nil {
		fmt.Println(colorize("Error reading input: "+err.Error(), red))
		os.Exit(1)
	}
	inputSize := len(data)

	result, err := decompress(data)
	if err != nil {
		fmt.Println(colorize("Decompression error: "+err.Error(), red))
		os.Exit(1)
	}
	outputSize := len(result)

	if *verbose {
		ratio := float64(outputSize) / float64(inputSize)
		if inputSize == 0 {
			ratio = 1
		}
		fmt.Println(colorize(fmt.Sprintf("Compressed size: %d bytes", inputSize), yellow))
		fmt.Println(colorize(fmt.Sprintf("Decompressed size: %d bytes", outputSize), yellow))
		fmt.Println(colorize(fmt.Sprintf("Expansion ratio: %.2fx", ratio), green))
	}

	if err := writeOutput(outputFile, result); err != nil {
		fmt.Println(colorize("Error writing output: "+err.Error(), red))
		os.Exit(1)
	}
	if outputFile != "" && outputFile != "-" {
		fmt.Println(colorize("Result written to "+outputFile, green))
	}
}

File: rle_decompress.cpp
// rle_decompress.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <cctype>
#include <stdexcept>
#include <vector>

using namespace std;

const string RESET = "\033[0m";
const string GREEN = "\033[92m";
const string RED = "\033[91m";
const string YELLOW = "\033[93m";

string colorize(const string& text, const string& color) {
    return color + text + RESET;
}

const char ESCAPE = '\\';

string decompress(const string& text) {
    if (text.empty()) return "";
    ostringstream result;
    size_t i = 0, n = text.size();
    while (i < n) {
        char ch = text[i];
        if (ch == ESCAPE) {
            if (i + 1 < n && text[i + 1] == ESCAPE) {
                result << ESCAPE;
                i += 2;
                continue;
            }
            if (i + 1 >= n) throw runtime_error("Unexpected end after escape");
            char repeat_char = text[i + 1];
            i += 2;
            string num_str;
            while (i < n && isdigit(text[i])) {
                num_str += text[i];
                ++i;
            }
            if (num_str.empty()) throw runtime_error("Missing number after escape");
            int count = stoi(num_str);
            result << string(count, repeat_char);
        } else {
            result << ch;
            ++i;
        }
    }
    return result.str();
}

string readFile(const string& filename) {
    if (filename == "-" || filename.empty()) {
        stringstream buffer;
        buffer << cin.rdbuf();
        return buffer.str();
    }
    ifstream f(filename);
    if (!f) throw runtime_error("Cannot open file: " + filename);
    stringstream buffer;
    buffer << f.rdbuf();
    return buffer.str();
}

void writeFile(const string& filename, const string& content) {
    if (filename == "-" || filename.empty()) {
        cout << content;
    } else {
        ofstream f(filename);
        if (!f) throw runtime_error("Cannot write file: " + filename);
        f << content;
    }
}

int main(int argc, char* argv[]) {
    vector<string> args;
    for (int i = 1; i < argc; ++i) args.push_back(argv[i]);
    if (args.empty()) {
        cout << colorize("Usage: rle_decompress <input> [output] [-v]", YELLOW) << endl;
        return 1;
    }
    string inputFile, outputFile;
    bool verbose = false;
    for (const string& arg : args) {
        if (arg == "-v" || arg == "--verbose") {
            verbose = true;
        } else if (inputFile.empty()) {
            inputFile = arg;
        } else {
            outputFile = arg;
        }
    }
    if (inputFile.empty()) {
        cout << colorize("Error: input file not specified", RED) << endl;
        return 1;
    }

    string data;
    try {
        data = readFile(inputFile);
    } catch (const exception& e) {
        cout << colorize("Error reading input: " + string(e.what()), RED) << endl;
        return 1;
    }
    size_t inputSize = data.size();

    string result;
    try {
        result = decompress(data);
    } catch (const exception& e) {
        cout << colorize("Decompression error: " + string(e.what()), RED) << endl;
        return 1;
    }
    size_t outputSize = result.size();

    if (verbose) {
        double ratio = (inputSize > 0) ? (double)outputSize / inputSize : 1.0;
        cout << colorize("Compressed size: " + to_string(inputSize) + " bytes", YELLOW) << endl;
        cout << colorize("Decompressed size: " + to_string(outputSize) + " bytes", YELLOW) << endl;
        cout << colorize("Expansion ratio: " + to_string(ratio) + "x", GREEN) << endl;
    }

    try {
        writeFile(outputFile, result);
        if (!outputFile.empty() && outputFile != "-") {
            cout << colorize("Result written to " + outputFile, GREEN) << endl;
        }
    } catch (const exception& e) {
        cout << colorize("Error writing output: " + string(e.what()), RED) << endl;
        return 1;
    }
    return 0;
}
